mesonbuild/compilers/mixins/clang.py

```diff
@@ -165,6 +165,12 @@ def get_lto_compile_args(self, *, threads: int = 0, mode: str = 'default') -> T.
             args.extend(super().get_lto_compile_args(threads=threads))
         return args
 
+    def linker_to_compiler_args(self, args: T.List[str]) -> T.List[str]:
+        if isinstance(self.linker, (ClangClDynamicLinker, MSVCDynamicLinker)):
+            return [flag if flag.startswith('-Wl,') else f'-Wl,{flag}' for flag in args]
```

Comment:
I think Meson has always assumed that linker arguments are already prefixed with `-Wl,`, see for example https://gitlab.freedesktop.org/gstreamer/gstreamer/-/blob/main/subprojects/gstreamer/meson.build#L86. That's probably because not all linker args can be prefixed that way? Surprisingly, this seems to not be even documented???? https://mesonbuild.com/Reference-manual_functions.html#add_project_link_arguments.

For consistency we should do it for all linkers or none... but I'm not sure why you do it only in this very specific case.

Reply:
> I think Meson has always assumed that linker arguments are already prefixed with `-Wl,`, see for example https://gitlab.freedesktop.org/gstreamer/gstreamer/-/blob/main/subprojects/gstreamer/meson.build#L86. That's probably because not all linker args can be prefixed that way?

Indeed, the meaning of `-Wl,` means that the compiler driver should not try to interpret it, but rather pass it unmodified to the linker -- which is very different from "flags that should be used when linking, not when compiling".

Examples where this fails would include `-fuse-ld=bfd`, which the compiler driver must interpret to decide which linker to execute and which linkers don't understand, or `-static` which acts differently from `-Wl,-static` as the former tells the compiler driver to compile a static binary, while the latter tells the linker alone to just load `*.a` archives and may be used to push/pop some static libraries and some shared ones.

Reply:
In the combination of Clang + LINK.EXE, assuming all flags are wrapped with `-Wl` is *not* correct; all linker flags, *including the ones internally supplied by Meson*, must be forwarded with `-Wl`.

```
-----------
Command line: `clang C:\Users\Amalia\Desktop\Rachel\bmpflash\build\meson-private\tmp1xjsk_eb\testfile.c -o C:\Users\Amalia\Desktop\Rachel\bmpflash\build\meson-private\tmp1xjsk_eb\output.exe -D_FILE_OFFSET_BITS=64 -O0 -Werror=implicit-function-declaration -WX -Wl,--add-stdcall-alias` -> 0
stdout:
LINK : warning LNK4044: unrecognized option '/-add-stdcall-alias'; ignored
-----------
stderr:
warning: unknown warning option '-WX' [-Wunknown-warning-option]
1 warning generated.
-----------
Compiler for C supports link arguments -Wl,--add-stdcall-alias: YES 
```

I did not notice, however, that in clang-cl we've got double nesting (and a broken linker detection-- see who it's calling for the linking process):

```
❯ py -3 E:\meson\meson.py setup build --wipe 
The Meson build system
Version: 1.2.99
Source dir: C:\Users\Amalia\Desktop\Rachel\bmpflash
Build dir: C:\Users\Amalia\Desktop\Rachel\bmpflash\build
Build type: native build
Project name: bmpflash
Project version: 0.0.1
Activating VS 17.8.0 Preview 1.0
C++ compiler for the host machine: clang-cl (clang-cl 16.0.5)
C++ linker for the host machine: lld-link lld-link 16.0.5
```

```
-----------
Command line: `clang-cl C:\Users\Amalia\Desktop\Rachel\bmpflash\build\meson-private\tmpwu6gdz3a\testfile.cpp /FeC:\Users\Amalia\Desktop\Rachel\bmpflash\build\meson-private\tmpwu6gdz3a\output.exe /nologo /showIncludes /utf-8 /MD /nologo /showIncludes /utf-8 /Od /Oi- /link -WX -guard:cf` -> 0
Compiler for C++ supports link arguments -guard:cf: YES 
Running compile:
Working directory:  C:\Users\Amalia\Desktop\Rachel\bmpflash\build\meson-private\tmpt3_2ydx4
Code:
 int main(void) { return 0; }

-----------
Command line: `clang-cl C:\Users\Amalia\Desktop\Rachel\bmpflash\build\meson-private\tmpt3_2ydx4\testfile.cpp /FeC:\Users\Amalia\Desktop\Rachel\bmpflash\build\meson-private\tmpt3_2ydx4\output.exe /nologo /showIncludes /utf-8 /MD /nologo /showIncludes /utf-8 /Od /Oi- /link -WX -Wl,-guard:cf` -> 4044
stdout:
LINK : warning LNK4044: unrecognized option '/Wl,-guard:cf'; ignored
LINK : error LNK1218: warning treated as error; no output file generated
-----------
stderr:
clang-cl: error: linker command failed with exit code 4044 (use -v to see invocation)
-----------
Compiler for C++ supports link arguments -Wl,-guard:cf: NO
```

Reply:
Sent a test to cover both cases, and fixed the linker preference forwarding for clang-cl.